docs/cardano-monads.rst

```diff
@@ -0,0 +1,772 @@
+
+The concrete implementations are defined separately from the class,
+taking the context as an explicit parameter::
+
+    --       Constraints necessary for the impl (can contain 'MonadIO' if the impl is prod-specific,
+    --             \                             but preferably shouldn't)
+    --              \
+    --               \          NodeDBs are a RocksDB-specific context
+    --                \                  /
+    dbGetRocks :: MonadRealDB ctx m => NodeDBs -> DBTag -> ByteString -> ReaderT ctx m (Maybe ByteString)
+    dbGetRocks = ...
+
+    -- Constraints for the impl (no 'MonadIO', because it's used for tests)
+    --           \
+    --            \               DBPureVar is a pure-db specific context
+    --             \                /
+    dbGetPure :: MonadPureDB m => DBPureVar -> DBTag -> ByteString -> ReaderT ctx m (Maybe ByteString)
+    dbGetPure = ...
+
+We can use these implementations to define instances for the base monads::
+
+    instance MonadDBGet Prod where
+        type DBContext Prod = NodeDBs
+        dbGet tag key = do
+          nodeDBs <- asks dbContext
+          dbGetRocks nodeDBs tag key
+
+    instance MonadDBGet Test where
+        type DBContext Test = DBPureVar
+        dbGet tag key = do
+          dbPureVar <- asks dbContext
+          dbGetPure dbPureVar tag key
+
+These instances are defined not in the modules with implementations, but in the modules
+with the base monads (``Prod`` and ``Test``), similarly to what we have now.
+
+The implementation-specific context is extracted in these method definitions.
+
+In case of ``MonadDBGet``, the choice of implementation is determined by the
+base monad (in prod we use rocksdb, in tests we use pure db). However, we also
+have cases where we need runtime configurability, when with the same base monad
+there can be different implementations. For instance, for slotting we can have
+ntp-based slotting or simple slotting, depending on whether the ``--no-ntp`` flag
+was passed. In these cases we choose to use effect sums::
+
+    instance MonadSlotting Prod where
+        type SlottingContext Prod = Either SimpleSlottingVar NtpSlottingVar
+        getCurrentSlot = do
+          sctx <- asks slottingContext
+          either getCurrentSlotSimple getCurrentSlotNtp sctx
+
+While effect context sums are closed, we are fine with it because we don't
+need extensibility at instance definition site, while the implementations
+themselves are defined independently.
+
+In order to address the issue of extensibility, for each ``Has``-class we can
+define two instances in a generic way::
+
+    class HasFooContext b ctx where
+        fooContext :: ctx -> FooContext b
+
+    instance HasFooContext b (FooContext b, _) where
+        fooContext = fst
+
+    instance {-# OVERLAPPABLE #-} HasFooContext b ctx => HasFooContext b (_, ctx) where
+        fooContext = fooContext . snd
+
+This way it's possible to add a context using ``withReader (fooCtx,)``. A
```

Comment:
Can you be more precise about your idea? I'm almost sure `withReader` is not exactly what you want to do here :)

Reply:
Oops, I meant `withReaderT` http://hackage.haskell.org/package/transformers-0.5.5.0/docs/Control-Monad-Trans-Reader.html#v:withReaderT